plugin/azure/azure.go

```diff
@@ -0,0 +1,227 @@
+					return
+				}
+				newZ := updateZoneFromResourceSet(recordSet, zName)
+				newZ.Upstream = h.upstream
+				h.zMu.Lock()
+				(*z[i]).z = newZ
+				h.zMu.Unlock()
+			}
+		}(zName, z)
+	}
+	// Collect errors (if any). This will also sync on all zones updates
+	// completion.
+	var errs []string
+	for i := 0; i < len(h.zones); i++ {
+		err := <-errc
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+	}
+	if len(errs) != 0 {
+		return fmt.Errorf("errors updating zones: %v", errs)
+	}
+	return nil
+}
+
+func updateZoneFromResourceSet(recordSet azure.RecordSetListResultPage, zName string) *file.Zone {
+	newZ := file.NewZone(zName, "")
+	for _, result := range *(recordSet.Response().Value) {
+		if result.RecordSetProperties.ARecords != nil {
+			for _, a := range *(result.RecordSetProperties.ARecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "A", *(a.Ipv4Address))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.AaaaRecords != nil {
+			for _, a := range *(result.RecordSetProperties.AaaaRecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "AAAA", *(a.Ipv6Address))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.NsRecords != nil {
+			for _, a := range *(result.RecordSetProperties.NsRecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "NS", *(a.Nsdname))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.SoaRecord != nil {
+			a := result.RecordSetProperties.SoaRecord
+			rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "SOA", fmt.Sprintf("%s %s %d %d %d %d %d", *(a.Host), *(a.Email), *(a.SerialNumber), *(a.RefreshTime), *(a.RetryTime), *(a.ExpireTime), *(a.MinimumTTL)))
+			r, err := dns.NewRR(rfc1035)
+			if err != nil {
+				log.Warningf("failed to process resource record set: %v", err.Error())
+			}
+			newZ.Insert(r)
+		}
+	}
+	return newZ
+}
+
+// ServeDNS uses the azure plugin to serve dns requests
+func (h *Azure) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+
+	zName := plugin.Zones(h.zoneNames).Matches(qname)
+	if zName == "" {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+
+	z, ok := h.zones[zName]
```

Comment:
when does this happen if we're already are authoritative for the zone? Please add a comment